src/meta/processors/admin/BalanceTask.cpp

```diff
@@ -51,10 +52,13 @@ void BalanceTask::invoke() {
             SAVE_STATE();
             if (srcLived_) {
                 client_->transLeader(spaceId_, partId_, src_).thenValue([this](auto&& resp) {
-                    if (!resp.ok()) {
-                        ret_ = Result::FAILED;
-                    } else {
-                        status_ = Status::ADD_PART_ON_DST;
+                    {
+                        std::lock_guard<std::mutex> lg(lock_);
```

Comment:
in folly::future's then function, it will call this call back directly in current thread stack when the future state is setResult, and this will dead lock, so need call via() function before thenValue.

Reply:
Good catch

Reply:
我记得client的线程池是初始化好的全局的io线程池，但是当前执行的线程是Balancer初始化的时候自己私有的线程池，所以会并行执行，可能会导致死锁。
Balancer(kvstore::KVStore* kv, std::unique_ptr<AdminClient> client)
        : kv_(kv)
        , client_(std::move(client)) {
        executor_.reset(new folly::CPUThreadPoolExecutor(1));
    }

Reply:
`getCPUExecutor` and `getIOExecutor` is the global thread pool. IMO, the `Balancer` is in a `CPUThreadPoolExecutor`, and all BalanceTask will be invoked in `IOThreadPoolExecutor` of `AdminClient`. So I guess the two pools have different queues and thread, we can't have dead locks here.

Reply:
因为是两个线程，所以调用client的函数的线程可能会提前结束，这样调用thenValue的时候，future的值已经存在了，所以会在当前线程栈中直接调用thenValue中的cb，这样就会死锁

Reply:
如果是同一个线程，就不会存在死锁的可能，因为在thenValue返回前，future中的值不可能ready

Reply:
`Deadlock` is when a set of processes are blocked because each process is holding a resource and waiting for another resource acquired by some other process.

Reply:
@darionyaphet 是的，还有就是同一个线程两次请求加同一个非递归锁也会死锁，这里会存在这种情况